### PR TITLE
ZCS-11122: Protect against XSS while rendering list view in /h/calendar

### PR DESCRIPTION
**Pull request by jholder85638**
When rendering the basic html client calendar view, protect against XSS using the fn:escapeXML function to render newAppt

### CODE REVIEW

File: WebRoot/WEB-INF/tags/calendar/monthView.tag [hunk 1]
```diff
@@ -124,7 +124,7 @@
                                 </c:when>
                                 <c:otherwise>
                                     <app:calendarUrl var="monthZoomUrl" view="month" zoom="${true}" timezone="${timezone}" rawdate="${currentDay}"/>
-                                    <td width="11%" onclick='javascript:dayClick(event, "${monthZoomUrl}");' class='ZhCalMonthDay${currentDay.timeInMillis eq date.timeInMillis ? 'Selected':''}'>
+                                    <td width="11%" onclick='javascript:dayClick(event, "${fn:escapeXml(monthZoomUrl)}");' class='ZhCalMonthDay${currentDay.timeInMillis eq date.timeInMillis ? 'Selected':''}'>
```
Comment (BhavinBathani) [inline WebRoot/WEB-INF/tags/calendar/monthView.tag:127]: I think we need to check if fn:escapeXml is proper function or zm:cook

Comment (silentsakky) [reply WebRoot/WEB-INF/tags/calendar/monthView.tag:127]: I see in many places in tags file we are using fn:escapeXml, so let's go ahead with that as a safer bet

**Review by silentsakky — APPROVED** [review]

**Review by k-kato — APPROVED** [review]

File: WebRoot/WEB-INF/tags/calendar/monthView.tag (hunk 1 above)
Comment (k-kato) [reply WebRoot/WEB-INF/tags/calendar/monthView.tag:127]: FYI:
* zm:cook  https://github.com/Zimbra/zm-mailbox/blob/develop/common/src/java/com/zimbra/common/util/StringUtil.java#L769
* fn:escapeXml  https://github.com/javaee/jstl-api/blob/master/impl/src/main/java/org/apache/taglibs/standard/tag/common/core/Util.java#L118

**Comment by silentsakky** [comment]
I am assuming this change will be applicable to 8.8.15 only, because in 9.0.0 we have removed /h/calendar endpoint as part of https://github.com/Zimbra/zm-web-client/pull/631 and https://github.com/Zimbra/zm-web-client/pull/628

File: WebRoot/WEB-INF/tags/calendar/monthView.tag (hunk 1 above)
Comment (jholder85638) [reply WebRoot/WEB-INF/tags/calendar/monthView.tag:127]: Generally, escapeHtml is considered much safer than escapeXml.
escapeXml only supports escaping the five basic XML entities (gt, lt, quot, amp, apos) whereas escapeHtml supports escaping all known [HTML 4.0 entities](http://www.w3.org/TR/REC-html40/sgml/entities.html).

In this instance, escapeXML was only chosen based on the previous code I saw.

Comment (k-kato) [reply WebRoot/WEB-INF/tags/calendar/monthView.tag:127]: @jholder85638 `zm:cook` or `com.zimbra.common.util.StringUtil#escapeHtml` removes the 5 characters plus CR and LF.

**Comment by jholder85638** [comment]
> I am assuming this change will be applicable to 8.8.15 only, because in 9.0.0 we have removed /h/calendar endpoint as part of #631 and #628

[These are my personal comments, and should not be used an official source of response from Synacor/Zimbra]
This is correct. ZCS 9.0 P15 or later cannot be affected. ZCS 9.0 P14 and earlier included some /h/ endpoints. A review of these by multiple people (myself included) did not find any vulnerabilities. 

Most Zimbra versions prior to 8.8.15 P30 are affected. Users should upgrade.

If users cannot upgrade, they must not apply this fix as it is only designed for 8.8.15. In that case, immediate triage can be achieved by removing /h/calendar altogether. This will break the basic HTML calendar, but does resolve the vulnerability in versions prior to 8.8.15

**Pull request merged**